File: gunhoflash/design-add-and-search-words-data-structure.ts
class WordDictionary {
  hasValue: boolean;
  subDictionary: Map<string, WordDictionary>;

  constructor() {
    this.hasValue = false;
    this.subDictionary = new Map<string, WordDictionary>();
  }

  addWord(word: string): void {
    if (word.length) {
      const firstLetter = word[0];
      const subDictionary = this.subDictionary.get(firstLetter) || new WordDictionary();
      subDictionary.addWord(word.slice(1));
      this.subDictionary.set(firstLetter, subDictionary);
    } else {
      this.hasValue = true;
    }
  }

  search(word: string): boolean {
    if (word.length) {
      const firstLetter = word[0];
      if (firstLetter === '.') {
        return Array.from(this.subDictionary).some(([, subDictionary]) => subDictionary.search(word.slice(1)));
      } else {
        return !!this.subDictionary.get(firstLetter)?.search(word.slice(1));
      }
    } else {
      return this.hasValue;
    }
  }
}

File: ezbroWorld/design-add-and-search-words-data-structure.py
from collections import defaultdict
class WordDictionary:

    def __init__(self):
        
        self.wordDict = defaultdict(list)
        

    def addWord(self, word: str) -> None:
        
        if len(word) in self.wordDict:
            self.wordDict[len(word)].append(word)
        else:
            self.wordDict[len(word)] = [word]
        

    def search(self, word: str) -> bool:
        
        if word == None:
            return False
        if '.' not in word:
            return word in self.wordDict[len(word)]
        for v in self.wordDict[len(word)]:
            for i, ch in enumerate(word):
                if v[i] != ch and ch != '.':
                    break
            else:
                return True
        return False
        
